### DISCUSSION TITLE
Difference between `SystemSet` and `Schedule`

### DISCUSSION THREAD

**Discussion by mamekoro** [bevyengine/bevy#9570] [Q&A] [answered]
(Bevy 0.11)

I want to create the following `SystemSet`s/`Schedule`s, and want to run them in `bevy::app::Update` in the following order.

1. `PreCollision` (runs systems to prepare resources for collision calculation)
2. `Collision` (runs systems to calculate collisions)
3. `PostCollision` (runs systems such as `despawn_destroyed_objects`)

However, I don't know what to do because I don't know the difference between `SystemSet` and `Schedule`.  What is the difference?

**Comment by jnhyatt** [comment, marked as answer]
Schedules are groups of systems with optional ordering constraints between them. If systems are the smallest building block for code execution, schedules are kind of the largest. Schedules run on the `World` and use their ordering constraints and run conditions to decide which of their systems to run and in what order. `Update` is one of the core Bevy schedules.

System sets live inside schedules and allow you to group systems as logical units, just like you're thinking here. You can put a set of systems into a system set and configure that entire set to run before another set, or configure that entire set to only run on a specific condition. 

You'll want to consider how your collision systems interact with the rest of the engine to decide which schedule they belong in, but Update is a very sane place to start. As an example, if your system needs to access transform data after the engine has run the transform propagation step, it'll need to go in PostUpdate because that's where `propagate_transforms` lives.

**Reply by mamekoro** [reply]
Now I understand. Thank you very much!

**Reply by lucidBrot** [reply]
@jnhyatt are there any example use-cases where you would prefer making a new custom Schedule instead of using a system set?

Would this be correct to say?:  
If I have some systems that must be logically ordered only against a few others, I can use system sets. If I have a distinct group of systems, that should always run without possibly being impacted by other systems running at the same time, I should use a custom Schedule.

**Reply by jnhyatt** [reply]
> @jnhyatt are there any example use-cases where you would prefer making a new custom Schedule instead of using a system set?
> 
> Would this be correct to say?:  
> If I have some systems that must be logically ordered only against a few others, I can use system sets. If I have a distinct group of systems, that should always run without possibly being impacted by other systems running at the same time, I should use a custom Schedule.

That's probably fair. Another use case I came up with, I was prototyping a management game where I wanted a bunch of systems to run once per day. The day loop is 100% decoupled from the main loop, meaning you can pause the in-game time progression completely, or you can run multiple days in one single frame. It's doable with system sets, but much much cleaner to create a Daily schedule (or in my case, I realized I can use the Fixed schedule as my day schedule, but not always applicable) that you can run on-demand. 

There's also been some interesting discussions I've seen about unifying system sets and schedules by making the scheduler smarter. It could open up a lot of really interesting spaces like async systems or completely blow away unnecessary hard sync lines between schedules.
